diff --git a/node_modules/bedrock-protocol/index.js b/node_modules/bedrock-protocol/index.js
index 7018ae3..1b38537 100644
--- a/node_modules/bedrock-protocol/index.js
+++ b/node_modules/bedrock-protocol/index.js
@@ -6,7 +6,7 @@ if (typeof process !== 'undefined' && parseInt(process.versions.node.split('.')[
 
 const { Client } = require('./src/client')
 const { Server } = require('./src/server')
-const { Relay } = require('./src/relay')
+const { Relay, RelayPlayer } = require('./src/relay')
 const { createClient, ping } = require('./src/createClient')
 const { createServer } = require('./src/createServer')
 const { Titles } = require('prismarine-auth')
@@ -16,6 +16,7 @@ module.exports = {
   Client,
   Server,
   Relay,
+  RelayPlayer,
   createClient,
   ping,
   createServer,
diff --git a/node_modules/bedrock-protocol/src/relay.js b/node_modules/bedrock-protocol/src/relay.js
index 3128a7d..251e385 100644
--- a/node_modules/bedrock-protocol/src/relay.js
+++ b/node_modules/bedrock-protocol/src/relay.js
@@ -47,6 +47,8 @@ class RelayPlayer extends Player {
     try {
       des = this.server.deserializer.parsePacketBuffer(packet)
     } catch (e) {
+      if (this.$hook$readUpstream$parsing_error?.(packet, e)) return
+
       this.server.deserializer.dumpFailedBuffer(packet, this.connection.address)
       console.error(this.connection.address, e)
 
@@ -60,6 +62,8 @@ class RelayPlayer extends Player {
     const params = des.data.params
     this.upInLog('->', name, params)
 
+    if (this.$hook$readUpstream$parsed?.(packet, des)) return
+
     if (name === 'play_status' && params.status === 'login_success') return // Already sent this, this needs to be sent ASAP or client will disconnect
 
     if (debugging) { // some packet encode/decode testing stuff
@@ -68,6 +72,8 @@ class RelayPlayer extends Player {
 
     this.emit('clientbound', des.data, des)
 
+    if (this.$hook$readUpstream$emitted?.(packet, des)) return
+
     if (!des.canceled) {
       if (name === 'start_game') {
         setTimeout(() => {
@@ -289,4 +295,4 @@ class Relay extends Server {
 }
 
 // Too many things called 'Proxy' ;)
-module.exports = { Relay }
+module.exports = { Relay, RelayPlayer }
